// wdio-testingbot-service/build/tb-launch-service.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _testingbotTunnelLauncher = require('testingbot-tunnel-launcher');

var _testingbotTunnelLauncher2 = _interopRequireDefault(_testingbotTunnelLauncher);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var TBLaunchService = function () {
    function TBLaunchService() {
        _classCallCheck(this, TBLaunchService);
    }

    _createClass(TBLaunchService, [{
        key: 'onPrepare',

        /**
         * modify config and launch tb tunnel
         */
        value: function onPrepare(config) {
            var _this = this;

            if (!config.tbTunnel) {
                return;
            }

            this.tbTunnelOpts = _extends({
                apiKey: config.user,
                apiSecret: config.key
            }, config.tbTunnelOpts);

            config.protocol = 'http';
            config.host = 'localhost';
            config.port = 4445;

            return new Promise(function (resolve, reject) {
                return (0, _testingbotTunnelLauncher2.default)(_this.tbTunnelOpts, function (err, tunnel) {
                    if (err) {
                        return reject(err);
                    }

                    _this.tunnel = tunnel;
                    resolve();
                });
            });
        }

        /**
         * shut down the tunnel
         */

    }, {
        key: 'onComplete',
        value: function onComplete() {
            var _this2 = this;

            if (!this.tunnel) {
                return;
            }

            return new Promise(function (resolve, reject) {
                return _this2.tunnel.close(resolve);
            });
        }
    }]);

    return TBLaunchService;
}();

exports.default = TBLaunchService;
module.exports = exports['default'];
