// wdio-testingbot-service/build/tb-service.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _request = require('request');

var _request2 = _interopRequireDefault(_request);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var jobDataProperties = ['name', 'tags', 'public', 'build', 'extra'];

var TBService = function () {
    function TBService() {
        _classCallCheck(this, TBService);
    }

    _createClass(TBService, [{
        key: 'before',

        /**
         * gather information about runner
         */
        value: function before(capabilities) {
            this.sessionId = global.browser.sessionId;
            this.capabilities = capabilities;
            this.auth = global.browser.requestHandler.auth || {};
            this.tbUser = this.auth.user;
            this.tbSecret = this.auth.pass;
            this.testCnt = 0;
            this.failures = 0;
        }
    }, {
        key: 'getRestUrl',
        value: function getRestUrl(sessionId) {
            return 'https://api.testingbot.com/v1/tests/' + sessionId;
        }
    }, {
        key: 'beforeSuite',
        value: function beforeSuite(suite) {
            this.suiteTitle = suite.title;
        }
    }, {
        key: 'beforeTest',
        value: function beforeTest(test) {
            if (!this.tbUser || !this.tbSecret) {
                return;
            }

            /**
             * in jasmine we get Jasmine__TopLevel__Suite as title since service using test
             * framework hooks in order to execute async functions.
             * This tweak allows us to set the real suite name for jasmine jobs.
             */
            if (this.suiteTitle === 'Jasmine__TopLevel__Suite') {
                this.suiteTitle = test.fullName.slice(0, test.fullName.indexOf(test.name) - 1);
            }

            global.browser.execute('tb:test-context=' + test.parent + ' - ' + test.title);
        }
    }, {
        key: 'afterTest',
        value: function afterTest(test) {
            if (!test.passed) {
                ++this.failures;
            }
        }
    }, {
        key: 'beforeFeature',
        value: function beforeFeature(feature) {
            if (!this.tbUser || !this.tbSecret) {
                return;
            }

            this.suiteTitle = feature.getName();
            global.browser.execute('tb:test-context=Feature: ' + this.suiteTitle);
        }
    }, {
        key: 'afterStep',
        value: function afterStep(feature) {
            if (feature.getFailureException()) {
                ++this.failures;
            }
        }
    }, {
        key: 'beforeScenario',
        value: function beforeScenario(scenario) {
            if (!this.tbUser || !this.tbSecret) {
                return;
            }

            global.browser.execute('tb:test-context=Scenario: ' + scenario.getName());
        }

        /**
         * update TestingBot info
         */

    }, {
        key: 'after',
        value: function after() {
            if (!this.tbUser || !this.tbSecret) {
                return;
            }

            return this.updateJob(this.sessionId, this.failures);
        }
    }, {
        key: 'onReload',
        value: function onReload(oldSessionId, newSessionId) {
            if (!this.tbUser || !this.tbSecret) {
                return;
            }

            this.sessionId = newSessionId;
            return this.updateJob(oldSessionId, this.failures, true);
        }
    }, {
        key: 'updateJob',
        value: function updateJob(sessionId, failures, calledOnReload) {
            var _this = this;

            return new Promise(function (resolve, reject) {
                return _request2.default.put(_this.getRestUrl(sessionId), {
                    json: true,
                    auth: {
                        user: _this.tbUser,
                        pass: _this.tbSecret
                    },
                    body: _this.getBody(failures, calledOnReload)
                }, function (e, res, body) {
                    if (e) {
                        return reject(e);
                    }
                    global.browser.jobData = body;
                    _this.failures = 0;
                    resolve(body);
                });
            });
        }

        /**
         * massage data
         */

    }, {
        key: 'getBody',
        value: function getBody(failures) {
            var calledOnReload = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;

            var body = { test: {} };

            /**
             * set default values
             */
            body.test['name'] = this.suiteTitle;

            /**
             * add reload count to title if reload is used
             */
            if (calledOnReload || this.testCnt) {
                body.test['name'] += ' (' + ++this.testCnt + ')';
            }

            var _iteratorNormalCompletion = true;
            var _didIteratorError = false;
            var _iteratorError = undefined;

            try {
                for (var _iterator = jobDataProperties[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
                    var prop = _step.value;

                    if (!this.capabilities[prop]) {
                        continue;
                    }

                    body.test[prop] = this.capabilities[prop];
                }
            } catch (err) {
                _didIteratorError = true;
                _iteratorError = err;
            } finally {
                try {
                    if (!_iteratorNormalCompletion && _iterator.return) {
                        _iterator.return();
                    }
                } finally {
                    if (_didIteratorError) {
                        throw _iteratorError;
                    }
                }
            }

            body.test['success'] = failures === 0 ? '1' : '0';
            return body;
        }
    }]);

    return TBService;
}();

exports.default = TBService;
module.exports = exports['default'];
